content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,352 @@
+```lua
+local alwaysHelloWorld: "Hello World!" = "Hello World!"
+local alwaysTrue: true = false -- TypeError: Type 'false' could not be converted into 'true'
+
+alwaysHelloWorld = "Goodbye World!" -- TypeError: Type '"Goodbyte World"' could not be into '"Hello World!"'
+```
+
+### Type Casts
+
+Sometimes, you may need to assist the typechecker by changing the type of a value using type casts, this can be done with the `::` operator.
+
+```lua
+local numericValue = 1
+
+local value = numericValue :: any -- ok, all expressions may be cast to 'any'
+local flag = numericValue :: boolean -- not ok, invalid 'number' to 'boolean' conversion
+```
+
+## Function Typing
+
+Assume the following function:
+
+```lua
+local function add(x, y)
+	return x + y
+end
+```
+
+This function will add x to y, however it might error if a string is given to this function, Luau doesn't know that this function can only use numbers, lets rework this function such that it does constrain `x` and `y` to be numbers.
+
+```lua
+local function add(x: number, y: number)
+	return x + y
+end
+```
+
+By adding type annotations to the parameters of the function, Luau now knows that the function takes two numbers and throws a warning if you try to pass anything that isn't a number into the function.
+
+```lua
+add(5, 10) -- ok
+add(5, "foo") -- TypeError: Type `string` could not be converted into `number`.
+```
+
+To define a return type, put a `:` operator at the end of the function definition.
+
+```lua
+local function add(x: number, y: number): number
+```
+
+To return multiple types, place all return types in brackets.
+
+```lua
+local function FindSource(script: BaseScript, pattern: string): (string, number)
+```
+
+### Defining a Functional Type
+
+A functional type can be defined by using the syntax `(in) -> out`. Using the functions from the previous examples, the types of the functions are:
+
+```lua
+type add = (x: number, y: number) -> number
+type FindSource = (script: BaseScript, pattern: string) -> (string, number)
+```
+
+## Table Types
+
+Luau does not have a `table` type, instead, table types are defined using `{}` syntax. One way of defining tables is using the `{type}` syntax, which defines a list type.
+
+```lua
+local numbers: {number} = {1, 2, 3, 4, 5}
+local characterParts: {Instance} = LocalPlayer.Character:GetChildren()
+```
+
+Index types can be defined using `{[indexType] = valueType}`.
+
+```lua
+local frobulator: {[string]: any} = {
+	Foo = "bar",
+	Baz = 10
+}
+```
+
+Tables can also have explicit string indices defined in a type.
+
+```lua
+type Car = {
+	Speed: number,
+	Drive: (Car) -> ()
+}
+```
+
+## Variadics
+
+Here's a function that calculates the sum of an arbitrary amount of numbers.
+
+```lua
+local function addLotsOfNumbers(...)
+	local sum = 0
+
+	for _, v in {...} do
+		sum += v
+	end
+
+	return sum
+end
+```
+
+As expected, this can take any value, and the typechecker wont raise a warning if you give it an invalid type, such as a `string`.

```

Comment:
```suggestion
As expected, this can take any value, and the typechecker won't raise a warning if you give it an invalid type, such as a `string`.

```